java/ql/lib/semmle/code/java/frameworks/apache/IOGenerated.qll

```diff
@@ -1,684 +1,828 @@
+        "org.apache.commons.io;IOUtils;true;toString;(InputStream,Charset);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;toString;(InputStream,String);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;toString;(Reader);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;toString;(URI);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;toString;(URI,Charset);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;toString;(URI,String);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;toString;(URL);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;toString;(URL,Charset);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;toString;(URL,String);;Argument[0];generated:open-url",
+        "org.apache.commons.io;IOUtils;true;write;(CharSequence,Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(String,Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(StringBuffer,Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(byte[],Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(byte[],Writer,Charset);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(byte[],Writer,String);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;write;(char[],Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;writeChunked;(char[],Writer);;Argument[0];generated:write-file",
+        "org.apache.commons.io;IOUtils;true;writeLines;(Collection,String,Writer);;Argument[1];generated:write-file",
+        "org.apache.commons.io;RandomAccessFileMode;false;create;(File);;Argument[0];generated:create-file",
+        "org.apache.commons.io;RandomAccessFileMode;false;create;(Path);;Argument[0];generated:create-file",
+        "org.apache.commons.io;RandomAccessFileMode;false;create;(String);;Argument[0];generated:create-file",
+        "org.apache.commons.io;UncheckedIO;true;apply;(IOFunction,Object);;Argument[1];generated:write-file",
+        "org.apache.commons.io;UncheckedIO;true;apply;(IOTriFunction,Object,Object,Object);;Argument[1];generated:write-file"
       ]
   }
 }
 
-private class IOSourcesCsv extends SourceModelCsv {
+private class IOGeneratedSourcesCsv extends SourceModelCsv {
   override predicate row(string row) {
     row =
       [
-        "org.apache.commons.io;IOUtils;true;resourceToByteArray;(String,ClassLoader);;ReturnValue;remote",
-        "org.apache.commons.io;IOUtils;true;toByteArray;(URI);;ReturnValue;remote"
+        "org.apache.commons.io.function;IOConsumer;true;forEachIndexed;(Stream,IOConsumer);;Argument[0];generated:remote",
```

Comment:
I'm not convinced these should be considered remote sources — at most they seem local (they're reading local files/resources). Are these generated this way because they return a remote source?

Reply:
I tried to make a path problem version of the create source model queries and ran it against commons-io.
It turns out that a path can be found, where the IOConsumer writes from a remote source to the Stream argument of forEachIndexed.

Reply:
What's the source? The method `IOStream.forEachIndexed` (which is called from `IOConsumer.forEachIndexed`) doesn't seem to contain any: https://github.com/apache/commons-io/blob/bcd5c9ba0cd1607b286ef0c116f119983e3f8a15/src/main/java/org/apache/commons/io/function/IOStreams.java#L39

Or is it that it exists one remote source somewhere in `commons-io` that gets passed to the first argument of `forEachIndexed`? If that's the case, it doesn't mean that the first argument of `forEachIndexed` is always a source, does it?

Reply:
Here is a screenshot of the path. The interesting bit is the jump from step 3-4. Not sure, how that happens. Basically, the entire method is considered a new remote due to this jump as inputstream is created from the URL Connection.

<img width="751" alt="image" src="https://user-images.githubusercontent.com/6472811/170258496-6fc11721-8700-4031-ac68-1901dc78ddb2.png">

Reply:
Should we consider to just delete this row from the generated file?

Reply:
This looks like virtual dispatch creating more data flow edges than it should. I think we should remove the whole `IOGeneratedSourcesCsv` class for the time being.